All project code/index.js

```diff
@@ -79,32 +86,58 @@ app.get("/login", (req, res) => {
 });
 
 // Handles user login from a /login POST request.
-app.post("/login", (req, res) => {
+app.post("/login", async (req, res) => {
     const username = req.body.username;
     const password = req.body.password;
-
     const query = "SELECT * FROM users WHERE username = $1";
 
-    db.any(query, [username])
-    .then(async (user) => {
-        const match = await bcrypt.compare(password, user[0].password);
-
-        if (match)
-        {
-            req.session.user = "TEST";
-            req.session.save();
-            res.redirect("/main");
-        }
-        else
-        {
-            console.log("Incorrect username or password.");
-            res.redirect("/login");
-        }
+    let user; // The user found (or not found) in the database
+
+    // Queries database to find matching user for the inputted username.
+    await db.any(query, [username])
+    .then((foundUser) => {
+        user = foundUser[0];
     })
     .catch((err) => {
-        console.log(err.message);
+        // Something went wrong, refresh the page for the user to try again.
         res.redirect("/login");
+        return console.log(err.message);
     });
+
+    const incorrect_login_feedback = "Incorrect username or password."
+
+    if (user == null) {
+        // User not found in database, so username does not exist.
+        console.log(incorrect_login_feedback);
+        res.render("pages/login", {
+            message: incorrect_login_feedback,
+            error: true
+        });
+
+        // early return so a password (that doesn't exist) isn't checked, and user isn't authenticated.
+        return; 
+    }
+
+    // Check if (encrypted) passwords match
+    const password_match = await bcrypt.compare(password, user.password);
+
+    if (password_match !== true) {
+        // Password is incorrect.
+        console.log(incorrect_login_feedback);
+        res.render("pages/login", {
+            message: incorrect_login_feedback,
+            error: true
+        });
+        // early return, so user isn't authenticated.
+        return;
+    }
+
+    // Authenticate user
+    req.session.user = "TEST";
```

Comment:
Why is this test? Shouldn't this be the user credentials received from the database?

Reply:
I have no idea, this was put in earlier (see deleted lines above). I just moved that line around a little.